### DISCUSSION TITLE
HEIF and color conversion

### DISCUSSION THREAD

**Discussion by pngunit** [ImageMagick/ImageMagick#2901] [Digital Image Processing]
I am very confused about color when converting to and from HEIF. This started because I was trying to convert a bunch of JPEGs to HEIF using `convert input.jpg output.heif` and the colors were consistently slightly off, at least when viewed on MacOS.

To narrow it down I created a png filled with a gray and converted it with `convert gray.png gray.heif`. Sure enough, the gray.png shows as #747474 on screen and the gray.heif shows as #747572.

I tried to get something reproducible:

```
$ convert -size 100x100 xc:lightgray -colorspace sRGB gray.png

$ identify gray.png
gray.png PNG 100x100 100x100+0+0 8-bit Gray 256c 312B 0.000u 0:00.000

$ convert gray.png -resize 1x1 txt:-
# ImageMagick pixel enumeration: 1,1,255,gray
0,0: (211)  #D3D3D3  gray(211)

$ convert gray.png -colorspace sRGB gray.heif

$ identify gray.heif
gray.heif HEIF 100x100 100x100+0+0 8-bit YCbCr 0.000u 0:00.000

$ convert -colorspace sRGB gray.heif -resize 1x1 txt:-
# ImageMagick pixel enumeration: 1,1,255,srgb
0,0: (211.701,210.815,210.114)  #D4D3D2  srgb(83.02%,82.6725%,82.3978%)

$ convert gray.heif gray2.png

$ identify gray2.png
gray2.png PNG 100x100 100x100+0+0 8-bit sRGB 2c 298B 0.000u 0:00.000

$ convert gray2.png -resize 1x1 txt:-
# ImageMagick pixel enumeration: 1,1,255,srgb
0,0: (212,211,210)  #D4D3D2  srgb(212,211,210)
```

Here it appears to me like the round trip png->heic->png results in a different image. I don't fully understand all the output though.

Can anyone explain to me what's going on and how I can get a consistent color reproduction when converting to and from HEIF?

**Comment by fmw42** [comment]
Check your profiles to be sure you have the correct ones and are using them consistently and properly converting to sRGB

**Reply by pngunit** [reply]
The commands I posted generate an image without a profile and do all processing on that. Can you elaborate on which one of them is wrong?

**Reply by fmw42** [reply]
This is not correct:

convert -colorspace sRGB gray.heif -resize 1x1 txt:-

The -colorspace should come after reading the raster heif file.

**Reply by pngunit** [reply]
Ok. It doesn't make any difference to the results:

```
$ convert -size 100x100 xc:lightgray -colorspace sRGB gray.png

$ identify gray.png
gray.png PNG 100x100 100x100+0+0 8-bit Gray 256c 312B 0.000u 0:00.000

$ convert gray.png -resize 1x1 txt:-
# ImageMagick pixel enumeration: 1,1,255,gray
0,0: (211)  #D3D3D3  gray(211)

$ convert gray.png -colorspace sRGB gray.heif

$ identify gray.heif
gray.heif HEIF 100x100 100x100+0+0 8-bit YCbCr 0.000u 0:00.000

$ convert gray.heif -colorspace sRGB -resize 1x1 txt:-
# ImageMagick pixel enumeration: 1,1,255,srgb
0,0: (211.701,210.815,210.114)  #D4D3D2  srgb(83.02%,82.6725%,82.3978%)

$ convert gray.heif gray2.png

$ identify gray2.png
gray2.png PNG 100x100 100x100+0+0 8-bit sRGB 2c 298B 0.000u 0:00.000

$ convert gray2.png -resize 1x1 txt:-
# ImageMagick pixel enumeration: 1,1,255,srgb
0,0: (212,211,210)  #D4D3D2  srgb(212,211,210)
```

**Comment by snibgo** [comment]
What version of IM? If you use Q8 non-HDRI then you can't expect better accuracy than one part in 256.

Your results are all 8-bit, and your results are accurate to one part in 256.

If you have at least Q16, I suggest you insert "-depth 16" before every output. Does that increase accuracy?

**Reply by pngunit** [reply]
> If you use Q8 non-HDRI 

Not sure what this means, but from the version output it looks like I'm using Q16 HDRI?

I'm using 7.0.10-45 on ARM, but I have seen the problem on x86 as well.

I don't see this problem converting to/from jpg, png or tiff. Only heif. This suggests to me there is a bug in libheif, or the YCbCr color space conversion (in 8 bit) is causing the inaccuracy like you suggest.

`convert` doesn't seem to pay any attention to `-depth 16` no matter where I put it; everything comes out as 8-bit.

```
$ convert --version
Version: ImageMagick 7.0.10-45 Q16 arm 2020-12-01 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5)
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib
```

**Comment by snibgo** [comment]
Yes, your "-version" says you have Q16 HDRI (floating-point).

But if HEIF files record nmbers as integers 0..255, then every pixel value is rounded to one of those, and we can't expect an accuracy much better than 1 in 256.

**Reply by pngunit** [reply]
Why is there any rounding occurring, when the input is an 8-bit PNG? And why doesn't this problem manifest when converting to other 8-bit image types?

**Comment by snibgo** [comment]
Your input is grayscale, with RGB channels. You convert that to HEIF with YCbCr channels. This is a fairly simple transformation, using a 3x3 floating-point matrix. The floating-point results of that are rounded to integers in the range 0..255.

I expect the JPG library recognises that your image is grayscale, and doesn't bother to calculate Cb or Cr. Any formats that record with RGB channels don't need any conversion.

**Reply by pngunit** [reply]
Ok thanks. So it sounds like YBcCr conversion is where the rounding error is creeping in, and this could be avoided by using RGB HEIC. However from this discussion it seems like that is not very well supported: https://github.com/strukturag/libheif/issues/62 I also tried using heif-enc as described there, but it produced garbage.

> I expect the JPG library recognises that your image is grayscale, and doesn't bother to calculate Cb or Cr. 

I don't think that's quite what's going on with the JPEG case, because I see basically the same thing with non-grayscale colors too.
